File: src/languages/welsh/welsh.ts
/**
 * Welsh (cy) phonemizer — canonical IPA, espeak-independent, Northern-leaning. Rule-based g2p (g2p.ts) +
 * PENULTIMATE stress + the Welsh vowel-length rule. A stressed monophthong in a long context — open, or before a
 * single voiced/fricative coda — takes full length (ː) in a monosyllable/final syllable (mis → miːs); in a penult
 * it stays SHORT and LAX (pobol → pɔbɔl, nesaf → nɛsav — the NW referee shows lax, not the espeak-tensed [o]/[e]).
 * Elsewhere it stays lax and short (bore → bɔrɛ). Diphthongs and circumflex vowels are already long and untouched.
 * See docs/investigations/cy_bringup_investigation.md.
 */
import type { Phonemizer } from "../../registry.ts";
import { makeSymbolNormalizer } from "../../core/normalizeSymbols.ts";
import { assembleClauses } from "../../core/clauses.ts";
import { loadTsvMap } from "../../core/loadTsv.ts";
import { type Seg, toSegments } from "./g2p.ts";
import { numberToWords } from "./numbers.ts";
import { normalizeWelsh } from "./normalize.ts";
import { MANIFEST } from "./manifest.ts";

// LEXICON (lexicon.tsv, kaikki/Wiktionary NW-derived): pronunciations the rules mis-derive — per-word ⟨ae⟩/⟨ai⟩
// diphthong QUALITY (aeres→eɨ), lexical ⟨y⟩-obscure/clear irregularities, loan vowels, monosyllable length. See
// tools/gen/build-cy-kaikki-dict.mts.
let LEX: Map<string, string> | undefined;
function lexicon(): Map<string, string> {
    if (LEX === undefined)
        LEX = loadTsvMap(import.meta.url, "lexicon.tsv", undefined, {
            optional: true,
        });
    return LEX;
}

const LONG = MANIFEST.longVowel; // lax short → long tense (adds ː): ɛ→eː, ɔ→oː, …
const LENGTHENS = new Set([...MANIFEST.lengthenBefore]); // single coda consonants that give a long/tense context

/** Apply the vowel-length rule to the stressed nucleus. `stress` is its index; `isFinal` = no nucleus follows. */
function applyLength(segs: Seg[], stress: number, isFinal: boolean): void {
    const v = segs[stress]!;
    if (v.long || v.ph === "ə") return; // diphthong / circumflex / schwa: never length-adjusted
    // coda = the consonants between this nucleus and the next nucleus (or word end)
    let coda = 0,
        single = "";
    for (let j = stress + 1; j < segs.length && !segs[j]!.nucleus; j++) {
        coda++;
        single = segs[j]!.ph;
    }
    const longContext = coda === 0 || (coda === 1 && LENGTHENS.has(single));
    if (!longContext) return; // lax + short (voiceless stop, m, ŋ, ɬ, cluster, or the deferred n/r/l)
    // Full length ː only in a monosyllable / final syllable; a PENULT keeps its short LAX quality (pobol→pɔbɔl,
    // nesaf→nɛsav — the NW referee shows lax; the espeak-tensed penult [o]/[e] was an oracle artifact, cf. i→ɨ).
    if (isFinal) v.ph = LONG[v.ph] ?? v.ph;
}

const EXCEPTIONS = MANIFEST.exceptions; // irregular function words (short/lax where the rule would lengthen)
const ENCLITICS = MANIFEST.enclitics; // apostrophe-contracted enclitics (o'r, hi'n)

/** One Welsh word → canonical IPA: scan, penultimate stress, vowel length. */
export function phonemizeWord(word: string): string {
    const lw = word.toLowerCase();
    const exc = EXCEPTIONS[lw.replace(/['’]/g, "")];
    if (exc !== undefined) return exc; // irregular function word (i → ɨ, bod → bɔd)
    const lex = lexicon().get(word) ?? lexicon().get(lw);
    if (lex !== undefined) return lex; // kaikki NW lexicon: rules-can't-derive words
    // Apostrophe enclitic (o'r → oːr, hi'n → hiːn): phonemize the STEM as its own word so its length rule sees the
    // real (open) syllable, then append the enclitic — instead of merging them into one closed syllable.
    const clitic = lw.match(/^(.+)['’]([a-z]+)$/);
    if (clitic && ENCLITICS[clitic[2]!]) return phonemizeWord(clitic[1]!) + ENCLITICS[clitic[2]!]!;
    const segs = toSegments(word.replace(/['’]/g, "")); // strip clitic apostrophes (mae'r, cymru'n)
    if (segs.length === 0) return "";
    const nucleiIdx = segs
        .map((s, i) => (s.nucleus ? i : -1))
        .filter((i) => i >= 0);
    if (nucleiIdx.length === 0) return segs.map((s) => s.ph).join("");
    // PENULTIMATE stress (the second-to-last nucleus; the only nucleus in a monosyllable).
    const stressN = nucleiIdx.length >= 2 ? nucleiIdx.length - 2 : 0;
    const stress = nucleiIdx[stressN]!;
    applyLength(segs, stress, stressN === nucleiIdx.length - 1);
    // NB: the letter ⟨i⟩ stays FRONT (i/ɪ/iː) everywhere — Northern Welsh centralizes only ⟨u⟩ and clear ⟨y⟩ to
    // ɨ, keeping the i/ɨ contrast (melin → mɛlɪn, gwin → ɡwiːn). Run 1's i→ɨ rules matched an espeak ARTEFACT the
    // independent NW referee contradicts, and were removed in Run 3. The residual is now purely n/r/l LENGTH.
    // Secondary stress on the first syllable when the primary is the 3rd nucleus or later (cymdeithasol →
    // ˌkəmdəᶦˈθasɔl).
    // (Final unstressed ⟨e⟩→[a] — bore→bɔra, carreg→karaɡ — is a colloquial NW reduction, but not reliably
    // rule-based: at corpus scale it net-regresses [-3.5%], too many final ⟨e⟩ stay ɛ. Left as a lexical residual.)
    const secondary = stressN >= 2 ? nucleiIdx[0]! : -1;
    let out = "";
    for (let i = 0; i < segs.length; i++) {
        if (i === stress) out += "ˈ";
        else if (i === secondary) out += "ˌ";
        // Degeminate: a written double consonant (nn, rr, …) is pronounced SINGLE (gorffennaf→ɡɔrfɛnav, torri→tɔrɪ).
        // It marks the preceding vowel short — applyLength already saw the doubled coda above, so only the OUTPUT
        // collapses. (ll/dd/ff/… are single digraph phonemes, not identical-adjacent, so untouched.)
        const s = segs[i]!;
        const prev = segs[i - 1];
        if (i > 0 && !s.nucleus && prev && !prev.nucleus && s.ph === prev.ph)
            continue;
        out += s.ph;
    }
    return out;
}

const CLAUSE_MARK = MANIFEST.clausePunctuation;
// Welsh groups thousands with COMMAS (1,400 — the TOKEN swallows the comma so the tier can still see the
// number); the dot is a DECIMAL (2.3 → "dau pwynt tri") or a version (802.11n), claimed by normalize.
const TOKEN =
    /([a-zâêîôûŵŷàèìòùïëöäüA-ZÂÊÎÔÛŴŶ]+(?:['’-][a-zâêîôûŵŷA-Z]+)*)|(\d{1,3}(?:,\d{3})+(?:\.\d+)?|\d+\.\d+|\d+)|([.!?…,;:])/gu;

// #562 symbol normalization — Welsh: "y cant" after the number (40 y cant, the BBC Cymru convention);
// nouns stay SINGULAR after numerals in Welsh, so one form suffices (deg doler, not *doleri*).
// cant/doler/punt/cilogram are referee-attested; cilometr/milimetr/centimetr are the standard
// borrowings, read by rule. `m` (metre) added for the corpus's 100m/230m running events and 4892 m.
const SYMBOLS = makeSymbolNormalizer({
    percent: ["y cant"],
    currency: { "$": ["doler"], "£": ["punt"], "¥": ["yen"] },
    units: { km: ["cilometr"], kg: ["cilogram"], mm: ["milimetr"], cm: ["centimetr"], m: ["metr"] },
});

class WelshPhonemizer implements Phonemizer {
    text(input: string): string {
        // normalize.ts FIRST, then the shared symbol tier — normalize's ordinal/era/version steps need
        // the number and its suffix still adjacent, which the tier would break.
        return assembleClauses(SYMBOLS(normalizeWelsh(input)), TOKEN, (m, sink) => {
            if (m[1]) sink.emit(phonemizeWord(m[1]));
            else if (m[2]) {
                const n = Number(m[2].replace(/,/gu, ""));
                for (const wd of numberToWords(n).split(" "))
                    sink.emit(phonemizeWord(wd));
            }
            else if (m[3]) {
                const mk = CLAUSE_MARK[m[3]];
                if (mk) sink.pause(mk);
            }
        });
    }
}

/** Build the Welsh phonemizer (rule-based; penultimate stress + vowel length). */
export function createWelsh(): Phonemizer {
    return new WelshPhonemizer();
}

File: csharp/Vernacula.Phonemizer/Languages/Welsh/Welsh.cs
/**
 * Welsh (cy) phonemizer — canonical IPA, Northern-leaning. Rule-based g2p (G2p.cs) + PENULTIMATE stress +
 * the Welsh vowel-length rule. A stressed monophthong in a long context — open, or before a single
 * voiced/fricative coda — takes full length (ː) in a monosyllable/final syllable (mis → miːs); in a penult
 * it stays SHORT and LAX (pobol → pɔbɔl, nesaf → nɛsav).
 *
 * Ported from src/languages/welsh/welsh.ts — see that file for the corpus evidence.
 */
using System.Text;
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Welsh;

public sealed class WelshPhonemizer : ILanguage
{
    // LEXICON (lexicon.tsv, kaikki/Wiktionary NW-derived): pronunciations the rules mis-derive — per-word
    // ⟨ae⟩/⟨ai⟩ diphthong QUALITY (aeres→eɨ), lexical ⟨y⟩-obscure/clear irregularities, loan vowels,
    // monosyllable length. Lazily loaded so merely importing this module does not parse the TSV.
    private static IReadOnlyDictionary<string, string>? LEX;
    private static IReadOnlyDictionary<string, string> Lexicon() =>
        LEX ??= LoadTsv.LoadTsvMap("languages/welsh", "lexicon.tsv", optional: true);

    private static readonly IReadOnlyDictionary<string, string> LONG = Manifest.MANIFEST.LongVowel;
    private static readonly IReadOnlySet<string> LENGTHENS =
        new HashSet<string>(Manifest.MANIFEST.LengthenBefore.Select(c => c.ToString()), StringComparer.Ordinal);

    private static readonly IReadOnlyDictionary<string, string> EXCEPTIONS = Manifest.MANIFEST.Exceptions;
    private static readonly IReadOnlyDictionary<string, string> ENCLITICS = Manifest.MANIFEST.Enclitics;
    private static readonly IReadOnlyDictionary<string, string> CLAUSE_MARK = Manifest.MANIFEST.ClausePunctuation;

    private static readonly JsRe APOSTROPHE = JsRegex.Compile("['’]", "g");
    private static readonly JsRe CLITIC = JsRegex.Compile("^(.+)['’]([a-z]+)$", "gu");
    private static readonly JsRe TOKEN_COMMAS = JsRegex.Compile(",", "gu");

    /** Apply the vowel-length rule to the stressed nucleus. `stress` is its index; `isFinal` = no nucleus follows. */
    private static void ApplyLength(List<Seg> segs, int stress, bool isFinal)
    {
        var v = segs[stress];
        if (v.Long || v.Ph == "ə") return; // diphthong / circumflex / schwa: never length-adjusted
        // coda = the consonants between this nucleus and the next nucleus (or word end)
        var coda = 0;
        var single = "";
        for (var j = stress + 1; j < segs.Count && !segs[j].Nucleus; j++)
        {
            coda++;
            single = segs[j].Ph;
        }
        var longContext = coda == 0 || (coda == 1 && LENGTHENS.Contains(single));
        if (!longContext) return; // lax + short (voiceless stop, m, ŋ, ɬ, cluster, or the deferred n/r/l)
        // Full length ː only in a monosyllable / final syllable; a PENULT keeps its short LAX quality.
        if (isFinal) v.Ph = LONG.TryGetValue(v.Ph, out var l) ? l : v.Ph;
    }

    /** One Welsh word → canonical IPA: scan, penultimate stress, vowel length. */
    public static string PhonemizeWord(string word)
    {
        var lw = Js.ToLowerCase(word);
        if (EXCEPTIONS.TryGetValue(APOSTROPHE.Replace(lw, ""), out var exc)) return exc;
        var lex = Lexicon().TryGetValue(word, out var l1) ? l1
            : Lexicon().TryGetValue(lw, out var l2) ? l2 : null;
        if (lex is not null) return lex;
        // Apostrophe enclitic (o'r → oːr, hi'n → hiːn): phonemize the STEM as its own word so its length rule
        // sees the real (open) syllable, then append the enclitic.
        var clitic = CLITIC.Match(lw);
        if (clitic.Success && ENCLITICS.TryGetValue(clitic.Groups[2].Value, out var encl))
            return PhonemizeWord(clitic.Groups[1].Value) + encl;
        var segs = G2p.ToSegments(APOSTROPHE.Replace(word, "")); // strip clitic apostrophes (mae'r, cymru'n)
        if (segs.Count == 0) return "";
        var nucleiIdx = new List<int>();
        for (var i = 0; i < segs.Count; i++) if (segs[i].Nucleus) nucleiIdx.Add(i);
        if (nucleiIdx.Count == 0) return string.Concat(segs.Select(s => s.Ph));
        // PENULTIMATE stress (the second-to-last nucleus; the only nucleus in a monosyllable).
        var stressN = nucleiIdx.Count >= 2 ? nucleiIdx.Count - 2 : 0;
        var stress = nucleiIdx[stressN];
        ApplyLength(segs, stress, stressN == nucleiIdx.Count - 1);
        // NB: the letter ⟨i⟩ stays FRONT (i/ɪ/iː) everywhere — Northern Welsh centralizes only ⟨u⟩ and clear
        // ⟨y⟩ to ɨ, keeping the i/ɨ contrast (melin → mɛlɪn, gwin → ɡwiːn).
        var secondary = stressN >= 2 ? nucleiIdx[0] : -1;
        var outp = new StringBuilder();
        for (var i = 0; i < segs.Count; i++)
        {
            if (i == stress) outp.Append('ˈ');
            else if (i == secondary) outp.Append('ˌ');
            // Degeminate: a written double consonant (nn, rr, …) is pronounced SINGLE. It marks the preceding
            // vowel short — ApplyLength already saw the doubled coda, so only the OUTPUT collapses.
            // (ll/dd/ff/… are single digraph phonemes, not identical-adjacent, so untouched.)
            var s = segs[i];
            var prev = i > 0 ? segs[i - 1] : null;
            if (i > 0 && !s.Nucleus && prev is not null && !prev.Nucleus && s.Ph == prev.Ph)
                continue;
            outp.Append(s.Ph);
        }
        return outp.ToString();
    }

    // Welsh groups thousands with COMMAS (1,400 — the TOKEN swallows the comma so the tier can still see the
    // number); the dot is a DECIMAL or a version, claimed by Normalize.
    private static readonly JsRe TOKEN = JsRegex.Compile(
        "(" + HostWord.HostWordRun(new[] { "Latin" }, "", "'’-") + ")|([1-9]\\d{0,2}(?:,\\d{3})+(?:\\.\\d+)?|\\d+\\.\\d+|\\d+)|([.!?…,;:])", "gu");

    /**
     * This language's OWN inventory. ⚠ TWO DIFFERENT QUESTIONS, KEPT APART: the TOKEN class above decides
     * where the SCRIPT boundary falls (routing), while this one decides whether the g2p has rules for these
     * letters. A token this class REJECTS carries a letter the language does not use — i.e. a foreign name.
     */
    private const string NATIVE_CLASS = "[a-zâêîôûŵŷàèìòùïëöäüA-ZÂÊÎÔÛŴŶÀÈÌÒÙÏËÖÄÜ]";
    /** ⚠ Mirrors the TS's exported `nat`, which test/lexicon-reachability.test.ts uses to assert that every
     *  key in this engine's lexicons survives its own fold. */
    public static readonly Func<string, string> Nat = HostWord.MakeNativiser(NATIVE_CLASS, "u");

    // symbol normalization — Welsh: "y cant" after the number (40 y cant, the BBC Cymru convention); nouns
    // stay SINGULAR after numerals in Welsh, so one form suffices (deg doler, not *doleri*).
    private static readonly Func<string, string> SYMBOLS = NormalizeSymbols.MakeSymbolNormalizer(new SymbolData
    {
        // `multiply` — the word is this language's OWN, harvested from its existing `×` rule. Declaring it
        // HERE is what makes ASCII `x` read like `×`. One word, so `by` is omitted and defaults to it — this
        // language does not split dimension from product.
        Multiply = new MultiplyDef { Times = "gwaith" },
        Percent = new[] { "y cant" },
        Currency = new Dictionary<string, IReadOnlyList<string>>
        {
            ["$"] = new[] { "doler" }, ["£"] = new[] { "punt" }, ["¥"] = new[] { "yen" },
        },
        Units = new Dictionary<string, IReadOnlyList<string>>
        {
            ["km"] = new[] { "cilomedr" }, ["kg"] = new[] { "cilogram" }, ["mm"] = new[] { "milimetr" },
            ["cm"] = new[] { "centimetr" }, ["m"] = new[] { "metr" },
        },
        // `cilomedr`, NOT `cilometr` — the corpus's own squared reading is `cilomedr sgwâr`, so the head
        // noun has to be that word or the tier composes half an attested collocation.
        ExponentWords = new ExponentWordsDef
        {
            Squared = new[] { "sgwâr" },
            Cubed = new[] { "ciwbig" },
            Position = ExponentPosition.After,
        },
    });

    public string Text(string input) =>
        // Normalize.cs FIRST, then the shared symbol tier — Normalize's ordinal/era/version steps need the
        // number and its suffix still adjacent, which the tier would break.
        Clauses.AssembleClauses(SYMBOLS(Normalize.NormalizeWelsh(input)), TOKEN, (m, sink) =>
        {
            if (m.Groups[1].Success && m.Groups[1].Value.Length > 0)
                sink.Emit(PhonemizeWord(Nat(m.Groups[1].Value)));
            else if (m.Groups[2].Success && m.Groups[2].Value.Length > 0)
            {
                // ⚠ THE STRIPPED STRING IS PASSED AS `raw` (#1095) — the fallback exists precisely because
                // the double cannot carry the digits, so it must not re-derive them from one.
                var digits = TOKEN_COMMAS.Replace(m.Groups[2].Value, "");
                foreach (var wd in Numbers.NumberToWords(Js.Number(digits), digits).Split(' '))
                    sink.Emit(PhonemizeWord(wd));
            }
            else if (m.Groups[3].Success && m.Groups[3].Value.Length > 0)
            {
                if (CLAUSE_MARK.TryGetValue(m.Groups[3].Value, out var mk) && mk.Length > 0) sink.Pause(mk);
            }
        });

    /** Build the Welsh phonemizer (rule-based; penultimate stress + vowel length). */
    public static ILanguage CreateWelsh() => new WelshPhonemizer();

    internal static void RegisterSelf() => Registry.Register("welsh", CreateWelsh);
}
